### PR TITLE
Implement differentiable WaterbalanceFD

### LINKED ISSUE #116: [Task] Make wofost_72_WLP_CWB differentiable and vectorized
After #107 , the model [config of Wofost72_WLP_CWB](https://github.com/ajwdewit/pcse/blob/master/pcse/conf/Wofost72_WLP_CWB.conf) consists:

SOIL = WaterbalanceFD
CROP = Wofost72

Wofost72 is already differentiable. [`WaterbalanceFD`](https://github.com/ajwdewit/pcse/blob/4d9f0e4f542e9062db338aaf1a227a75f1b03949/pcse/soil/classic_waterbalance.py#L101) needs to be added to `classic_waterbalance.py` script in diffwofost and be made differentiable. Similar to other models, test should be added to show both differentiablilty and vectorization works. No need to add a notebook here.

### PR DESCRIPTION
**Pull request by SCiarella**
Closes #116.

This PR introduces a differentiable and vectorized version of `WaterbalanceFD` in diffwofost.

The implementation is tested similarly to `WaterbalancePP`, so there are some specific unit tests for the class, the differentiability and vectorization is tested, and also the integration in wofost72. I am not sure if specific data for `WaterbalanceFD` exist similarly to the `waterlimited_production`, so the numerical results are not compared against `pcse`.

### CODE REVIEW

File: src/diffwofost/physical_models/soil/classic_waterbalance.py [hunk 1]
```diff
@@ -101,3 +110,563 @@ def integrate(self, day, delt=1.0):
...
+        DSS = Tensor(0.0)
+        DRAINT = Tensor(0.0)
+
+    def initialize(self, day, kiosk, parvalues, shape=None):
+        """Initialize the freely-draining water balance.
+
+        :param day: start date of the simulation
+        :param kiosk: variable kiosk of this PCSE instance
+        :param parvalues: ParameterProvider containing all parameters
+        :param shape: Optional tensor shape for batched simulation
```
Comment (SarahAlidoost) [inline src/diffwofost/physical_models/soil/classic_waterbalance.py]:
```suggestion

        Args:
            day: Start date of the simulation.
            kiosk: Variable kiosk used to read and publish crop state.
            parvalues: Parameter provider containing the physical-model
                parameters for the soil.
            shape: Target tensor shape for state and rate variables.
```

Comment (SarahAlidoost) [reply src/diffwofost/physical_models/soil/classic_waterbalance.py]: can you please fix the format of docstring in `WaterbalancePP`, too?

File: src/diffwofost/physical_models/soil/classic_waterbalance.py [hunk 2]
```diff
@@ -101,3 +110,563 @@ def integrate(self, day, delt=1.0):
...
+        device = ComputeConfig.get_device()
+
+        self.params = self.Parameters(parvalues, shape=shape)
+        p = self.params
+
+        # Check validity of SMLIM (site parameter, not stored in self.params)
+        SMLIM_raw = torch.as_tensor(parvalues["SMLIM"], dtype=dtype, device=device)
+        SMLIM = torch.maximum(p.SMW, torch.minimum(p.SM0, SMLIM_raw))
+        if torch.any(SMLIM != SMLIM_raw):
+            self.logger.warn("SMLIM not in valid range, adjusted.")
```
Comment (SarahAlidoost) [inline src/diffwofost/physical_models/soil/classic_waterbalance.py]: The method "warn" in class "Logger" is deprecated since Python 3.3. Can you please replace it with  `Logger.warning()` instead.

File: src/diffwofost/physical_models/soil/classic_waterbalance.py [hunk 3]
```diff
@@ -101,3 +110,563 @@ def integrate(self, day, delt=1.0):
...
+        :param day: start date of the simulation
+        :param kiosk: variable kiosk of this PCSE instance
+        :param parvalues: ParameterProvider containing all parameters
+        :param shape: Optional tensor shape for batched simulation
+        """
+        dtype = ComputeConfig.get_dtype()
+        device = ComputeConfig.get_device()
+
+        self.params = self.Parameters(parvalues, shape=shape)
+        p = self.params
```
Comment (SarahAlidoost) [inline src/diffwofost/physical_models/soil/classic_waterbalance.py]:
```suggestion
```

Comment (SarahAlidoost) [reply src/diffwofost/physical_models/soil/classic_waterbalance.py]: can you please replace `p` with `self.params` in the codes below in `initialize` function.

File: src/diffwofost/physical_models/soil/classic_waterbalance.py [hunk 4]
```diff
@@ -101,3 +110,563 @@ def integrate(self, day, delt=1.0):
...
+        SMLIM_raw = torch.as_tensor(parvalues["SMLIM"], dtype=dtype, device=device)
+        SMLIM = torch.maximum(p.SMW, torch.minimum(p.SM0, SMLIM_raw))
+        if torch.any(SMLIM != SMLIM_raw):
+            self.logger.warn("SMLIM not in valid range, adjusted.")
+
+        # Default rooting depth: 10 cm; derive maximum rootable depth
+        RD = torch.as_tensor(self.DEFAULT_RD, dtype=dtype, device=device)
+        RDM = torch.maximum(RD, p.RDMSOL)
+        self.RDold = RD
+        self.RDM = RDM
```
Comment (SarahAlidoost) [inline src/diffwofost/physical_models/soil/classic_waterbalance.py]:
```suggestion
        self.RDold = torch.as_tensor(self.DEFAULT_RD, dtype=dtype, device=device)
        self.RDM = torch.maximum(self.RDold, self.params.RDMSOL)
```

Comment (SarahAlidoost) [reply src/diffwofost/physical_models/soil/classic_waterbalance.py]: In the codes below, can you please replace `RD` with `self.RDold `and `RDM` with `self.RDM`

File: src/diffwofost/physical_models/soil/classic_waterbalance.py [hunk 5]
```diff
@@ -101,3 +110,563 @@ def integrate(self, day, delt=1.0):
...
+
+        # Initial surface storage
+        SS = p.SSI
+
```
Comment (SarahAlidoost) [inline src/diffwofost/physical_models/soil/classic_waterbalance.py]:
```suggestion
```

File: src/diffwofost/physical_models/soil/classic_waterbalance.py [hunk 6]
```diff
@@ -101,3 +110,563 @@ def integrate(self, day, delt=1.0):
...
+        self.rooted_layer_needs_reset = False
+        self.NINFTB = Afgen([0.0, 0.0, 0.5, 0.0, 1.5, 1.0])
+        self._increments_w = []
+
+        # Initialize state and rate variables
+        self.states = self.StateVariables(
+            kiosk,
+            publish=["SM", "DSOS"],
+            SM=SM,
+            SS=SS,
```
Comment (SarahAlidoost) [inline src/diffwofost/physical_models/soil/classic_waterbalance.py]:
```suggestion
            SS=self.params.SSI,
```

File: src/diffwofost/physical_models/soil/classic_waterbalance.py [hunk 7]
```diff
@@ -101,3 +110,563 @@ def integrate(self, day, delt=1.0):
...
+        self.RDold = RD
+        self.RDM = RDM
+
+        # Initial surface storage
+        SS = p.SSI
+
+        # Initial soil moisture content in rooted zone, clamped to [SMW, SMLIM]
+        SM = torch.clamp(p.SMW + p.WAV / RD, min=p.SMW, max=SMLIM)
+        W = SM * RD
+        WI = W
```
Comment (SarahAlidoost) [inline src/diffwofost/physical_models/soil/classic_waterbalance.py]:
```suggestion
```

File: src/diffwofost/physical_models/soil/classic_waterbalance.py [hunk 8]
```diff
@@ -101,3 +110,563 @@ def integrate(self, day, delt=1.0):
...
+
+        # Initialize state and rate variables
+        self.states = self.StateVariables(
+            kiosk,
+            publish=["SM", "DSOS"],
+            SM=SM,
+            SS=SS,
+            SSI=p.SSI,
+            W=W,
+            WI=WI,
```
Comment (SarahAlidoost) [inline src/diffwofost/physical_models/soil/classic_waterbalance.py]:
```suggestion
            WI=W,
```

File: src/diffwofost/physical_models/soil/classic_waterbalance.py [hunk 9]
```diff
@@ -101,3 +110,563 @@ def integrate(self, day, delt=1.0):
...
+        W = SM * RD
+        WI = W
+
+        # Initial water in subsoil (below rooted zone to maximum rootable depth)
+        WLOW = torch.clamp(
+            p.WAV + RDM * p.SMW - W,
+            min=torch.zeros_like(W),
+            max=p.SM0 * (RDM - RD),
+        )
+        WLOWI = WLOW
```
Comment (SarahAlidoost) [inline src/diffwofost/physical_models/soil/classic_waterbalance.py]:
```suggestion
```

File: src/diffwofost/physical_models/soil/classic_waterbalance.py [hunk 10]
```diff
@@ -101,3 +110,563 @@ def integrate(self, day, delt=1.0):
...
+        self.states = self.StateVariables(
+            kiosk,
+            publish=["SM", "DSOS"],
+            SM=SM,
+            SS=SS,
+            SSI=p.SSI,
+            W=W,
+            WI=WI,
+            WLOW=WLOW,
+            WLOWI=WLOWI,
```
Comment (SarahAlidoost) [inline src/diffwofost/physical_models/soil/classic_waterbalance.py]:
```suggestion
            WLOWI=WLOW,
```

File: src/diffwofost/physical_models/soil/classic_waterbalance.py [hunk 11]
```diff
@@ -101,3 +110,563 @@ def integrate(self, day, delt=1.0):
...
+            WWLOW=WWLOW,
```
Comment (SarahAlidoost) [inline src/diffwofost/physical_models/soil/classic_waterbalance.py]:
```suggestion
            WWLOW=(W + WLOW),  # Total water in soil column
```

File: src/diffwofost/physical_models/soil/classic_waterbalance.py [hunk 12]
```diff
@@ -101,3 +110,563 @@ def integrate(self, day, delt=1.0):
...
+        WLOW = torch.clamp(
+            p.WAV + RDM * p.SMW - W,
+            min=torch.zeros_like(W),
+            max=p.SM0 * (RDM - RD),
+        )
+        WLOWI = WLOW
+
+        # Total water in soil column
+        WWLOW = W + WLOW
+
```
Comment (SarahAlidoost) [inline src/diffwofost/physical_models/soil/classic_waterbalance.py]:
```suggestion
```

File: src/diffwofost/physical_models/soil/classic_waterbalance.py [hunk 13]
```diff
@@ -101,3 +110,563 @@ def integrate(self, day, delt=1.0):
...
+        )
+        self.rates = self.RateVariables(kiosk, publish="EVS", shape=shape)
+
+        # Connect signals
+        self._connect_signal(self._on_CROP_START, signals.crop_start)
+        self._connect_signal(self._on_CROP_FINISH, signals.crop_finish)
+        self._connect_signal(self._on_IRRIGATE, signals.irrigate)
+
+    def calc_rates(self, day, drv):
+        """Calculate the rates of change for all water balance components."""
```
Comment (SarahAlidoost) [inline src/diffwofost/physical_models/soil/classic_waterbalance.py]:
```suggestion
        """Calculate the rates of change for all water balance components.
        
        Args:
            day (datetime.date): The current date of the simulation.
            drv (WeatherDataContainer): A dictionary-like container holding
                weather data elements as key/value. The values are
                arrays or scalars. See PCSE documentation for details.
        
        """
```

File: src/diffwofost/physical_models/soil/classic_waterbalance.py [hunk 14]
```diff
@@ -101,3 +110,563 @@ def integrate(self, day, delt=1.0):
...
+        r.DW = r.RIN - r.WTRA - r.EVS - r.PERC
+        r.DWLOW = r.PERC - r.LOSS
+
+        # Prevent W from going negative by reducing EVS
+        Wtmp = s.W + r.DW
+        under_zero = Wtmp < 0.0
+        r.EVS = torch.where(under_zero, torch.maximum(torch.zeros_like(r.EVS), r.EVS + Wtmp), r.EVS)
+        r.DW = torch.where(under_zero, -s.W, r.DW)
+
+        # Surface storage and runoff
```
Comment (SarahAlidoost) [inline src/diffwofost/physical_models/soil/classic_waterbalance.py:522]:
```suggestion
        # Surface storage and runoff
        # SStmp is the layer of water that cannot infiltrate and that can potentially
        # be stored on the surface. Here we assume that RAIN_NOTINF automatically
        # ends up in the surface storage (and finally runoff).
```

File: src/diffwofost/physical_models/soil/classic_waterbalance.py [hunk 15]
```diff
@@ -101,3 +110,563 @@ def integrate(self, day, delt=1.0):
...
+        SStmp = RAIN_t + r.RIRR - r.EVW - r.RIN
+        r.DSS = torch.minimum(SStmp, p.SSMAX - s.SS)
```
Comment (SarahAlidoost) [inline src/diffwofost/physical_models/soil/classic_waterbalance.py:524]:
```suggestion
        # rate of change in surface storage is limited by SSMAX - SS
        r.DSS = torch.minimum(SStmp, p.SSMAX - s.SS)
```

File: src/diffwofost/physical_models/soil/classic_waterbalance.py [hunk 16]
```diff
@@ -101,3 +110,563 @@ def integrate(self, day, delt=1.0):
...
+        r.DTSR = SStmp - r.DSS
```
Comment (SarahAlidoost) [inline src/diffwofost/physical_models/soil/classic_waterbalance.py:525]:
```suggestion
        # Remaining part of SStmp is send to surface runoff
        r.DTSR = SStmp - r.DSS
```

File: src/diffwofost/physical_models/soil/classic_waterbalance.py [hunk 17]
```diff
@@ -101,3 +110,563 @@ def integrate(self, day, delt=1.0):
...
+        r.DRAINT = RAIN_t
```
Comment (SarahAlidoost) [inline src/diffwofost/physical_models/soil/classic_waterbalance.py:527]:
```suggestion
        # incoming rainfall rate
        r.DRAINT = RAIN_t
```

File: src/diffwofost/physical_models/soil/classic_waterbalance.py [hunk 18]
```diff
@@ -101,3 +110,563 @@ def integrate(self, day, delt=1.0):
...
+
+    def integrate(self, day, delt=1.0):
+        """Integrate state variables over one time step."""
```
Comment (SarahAlidoost) [inline src/diffwofost/physical_models/soil/classic_waterbalance.py]:
```suggestion
        """Integrate state variables over one time step.

        Args:
            day (datetime.date): The current date of the simulation.
            delt (float, optional): The time step for integration. Defaults to 1.0.
        """
```

File: src/diffwofost/physical_models/soil/classic_waterbalance.py [hunk 19]
```diff
@@ -101,3 +110,563 @@ def integrate(self, day, delt=1.0):
...
+        if "RD" in self.kiosk:
+            return self.kiosk["RD"]
+        dtype = ComputeConfig.get_dtype()
+        device = ComputeConfig.get_device()
+        return torch.as_tensor(self.DEFAULT_RD, dtype=dtype, device=device)
+
+    def _redistribute_water(self, RDchange):
+        """Redistribute water between root zone and lower zone after RD changes.
+
+        :param RDchange: Change in root depth (cm); positive = downward growth.
```
Comment (SarahAlidoost) [inline src/diffwofost/physical_models/soil/classic_waterbalance.py]:
```suggestion
        Redistribution of water is needed when roots grow during the growing season
        and when the crop is finished and the root zone shifts back from the crop rooted
        depth to the default depth of the upper (rooted) layer of the water balance.
        Or when the initial rooting depth of a crop is different from the default one used
        by the water balance module (10 cm)
        
        Args:
            RDchange: Change in root depth (cm); positive = downward growth.
```

File: tests/physical_models/crop/test_wofost72.py [hunk 20]
```diff
@@ -655,6 +662,70 @@ class MockKiosk(VariableKiosk):
         assert crop.lv_dynamics.payload is component_models["leaf_dynamics"]
         assert crop.lv_dynamics._label == "leaf_dynamics"
 
+    def test_wofost72_runs_with_waterbalance_fd(self, device):
```
Comment (SarahAlidoost) [inline tests/physical_models/crop/test_wofost72.py]: For this test, since we are evaluating the model results, we need to make sure the results stay the same for all test data (test_waterlimitedproduction_wofost72_*). Can you please add others too, similar to the test:
```
    @pytest.mark.parametrize("test_data_url", wofost72_data_urls)
    def test_wofost72_with_testengine(self, test_data_url, device)
```

To organize the tests, we can also rename  `class TestWofost72` to `class TestWofost72_PP` and define a new one `class TestWofost72_WLP` for new tests.

File: tests/physical_models/soil/test_waterbalance.py [hunk 21]
```diff
@@ -431,3 +432,360 @@ def test_gradients_numerical(self, param_name, output_name, config_type, device)
...
+    param_names = ["SMFCF", "SMW", "SM0"]
+    output_names = ["SM", "EVS"]
+
+    # All three soil moisture parameters influence both SM and EVS.
+    # EVS has a multi-timestep gradient path: param → RIN → DW → W → SM → EVSMX → EVS.
+    gradient_mapping = {
+        "SMFCF": ["SM", "EVS"],
+        "SMW": ["SM", "EVS"],
+        "SM0": ["SM", "EVS"],
+    }
```
Comment (SarahAlidoost) [inline tests/physical_models/soil/test_waterbalance.py:684]: What about other soil parameters like `CRAIRC`, `SOPE`, `KSUB`, `RDMSOL`, can we also check the gradients of these parameters?

**Review by SarahAlidoost — CHANGES_REQUESTED** [review]
@SCiarella thanks, looks great 👍 . I left some comments, let me know if something is unclear. Another thing, can you please add the API reference to the documentation too?

File: tests/physical_models/crop/test_wofost72.py (hunk 20 above)
Comment (SCiarella) [reply tests/physical_models/crop/test_wofost72.py]: Good idea! I have separated the classes and added all the tests that i missed

File: tests/physical_models/soil/test_waterbalance.py (hunk 21 above)
Comment (SCiarella) [reply tests/physical_models/soil/test_waterbalance.py:684]: Thanks for noticing this! It has been fixed

**Comment by SCiarella** [comment]
@SarahAlidoost Thanks for the review!

I have implemented your comments and updated tests and documentation.

**Review by SarahAlidoost — APPROVED** [review]
@SCiarella thanks for addressing the comments 👍 let's merge it.
